Docs: Updated wallets section with newly integrated wallets (#266)
rsksmart/devportal: tutorials/tokens/create-a-token.md

---
layout: rsk
title: Create a token
tags: tutorial, rsk, token, openzeppelin, erc20, truffle 
description: "How to create a token less than 10 lines of code, using Truffle and Open Zeppelin smart contracts at RSK testnet"
---

In this tutorial I will show you step-by-step how to create a token with less than 10 lines of code, using Truffle plus Open Zeppelin smart contracts, and deploy it to the RSK testnet.

# Overview

Here is a summary of the steps to be taken to build our front end:

1. Initialize a project using Truffle;
2. Install Open Zeppelin smart contracts at project;
3. Create a wallet mnemonic;
4. Configure Truffle to connect to RSK testnet;

> 5. Get some testnet R-BTCs at faucet;

Reviewer: ```suggestion
5. Get some tR-BTC from a faucet;
```

6. Create smart contract of token;
7. Create deploy file at Truffle;
8. Deploy a smart contract on RSK Testnet using Truffle;
9. Interact with the smart contract at Truffle console.

# Requirements

* Node.js and NPM (Node Package Manager)
* Visual Studio Code (VSCode) or any other editor of your choice
* Truffle

## Node.js and NPM

Node.js and NPM are needed, though both are usually installed at once.

NB: To check if Node.js and NPM is already installed, input the following commands in the terminal:

```shell
node --version
npm --version
```

![node and npm version](/assets/img/tutorials/create-a-token/image-01.png)

Go to [Node.js](https://nodejs.org/en/) if you need to install it.

## Visual Studio Code (VSCode)

> In this tutorial, We would use VSCode to create and deploy our project.

Reviewer: ```suggestion
In this tutorial, we will use VSCode to create and deploy our project.
Feel free to use any other code editor of your choice.
```

To use VSCode [download it here](https://code.visualstudio.com/download).

Verify if your VS code installation was successful by typing the following command into the terminal:

```shell
code -v
```

![visual code version](/assets/img/tutorials/create-a-token/image-02.png)

## Truffle

> Truffle is a popular development framework for Ethereum with a mission to make your life a whole lot easier. 

Reviewer: ```suggestion
Truffle is a popular development framework for smart contract developers with a mission to make your work a whole lot easier. 
```

Among its features, it has smart contract lifecycle management, scriptable deployment & migrations, automated contract testing and simple network management. 

It can make the RSK life easier too because we can configure a RSK network in Truffle.

> To install Truffle, input the command below into the terminal and press `enter` at your project location:

Reviewer: ```suggestion
To install Truffle, input the command below into the terminal and press `enter`, in the directory of your project:
```

# Verify that you can connect to RSK Testnet

Enter the following command into your terminal. If you are using a Windows OS, I suggest to use the Git Bash terminal.

```shell
curl https://public-node.testnet.rsk.co/1.3.0/ \
  -X POST -H "Content-Type: application/json" \
  --data '{"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}'
```

This is a very simple query that simply asks what the latest block number is.

You should receive a response similar to the following:

```json
{"jsonrpc":"2.0","id":1,"result":"0xc1266"}
```

If you have Windows OS and you do not have Git installed, download the installer for Windows from the [Git official site](https://git-scm.com/).

# Initialize a Truffle project

Create a new folder, in a folder named `token`.

Start an Truffle project in the token folder by typing the following commands below into the terminal:

```shell
mkdir token
cd token
truffle init
```

![truffle init](/assets/img/tutorials/create-a-token/image-05.png)

Open the folder in VSCode. 
Then you can see the file structure like this:

![truffle file structure](/assets/img/tutorials/create-a-token/image-06.png)

> * ./contracts: all our smart contracts will be stored in this folder.
> * ./migrations: deployment scripts will be stored in this folder.
> * ./test: test scripts will be stored in this folder.
> * ./truffle-config: this is Truffle’s configuration file. We’ll be able to configure networks, including RSK networks.
> 
> Note that files were also created:
> 
> * Migrations.sol - keeps track of which migrations were done on the current network.
> * 1_initial_migration.js - deploy Migrations.sol

Reviewer: ```suggestion
* `./contracts`: All our smart contracts will be stored in this folder.
* `./migrations`: Deployment scripts will be stored in this folder.
* `./test`: Test scripts will be stored in this folder.
* `./truffle-config.js`: This is Truffle’s configuration file. We’ll be able to configure networks, including RSK networks.

Note that the following files were also created:

* `Migrations.sol`: Keeps track of which migrations were done on the current network.
* `1_initial_migration.js`: Deployment for `Migrations.sol`.
```

## Initialize an npm project

Start an npm project in the token folder by typing the following commands below into the terminal:

```shell
npm init -y
```

![npm init](/assets/img/tutorials/create-a-token/image-07.png)

## Install Open Zeppelin

> OpenZeppelin Contracts is a set of libraries of smart contracts for Ethereum and other blockchains. 

Reviewer: ```suggestion
OpenZeppelin Contracts is a set of smart contract libraries for Ethereum. They work well with other compatible blockchains, including **RSK**. 
```

These libraries will install not only the main libraries of our token but also libraries for ownership, safe math, and many other utilities. 

> It’s worth mentioning that these libraries have been reviewed and audited to accomplish high standards of security so contracts that depend on them are less susceptible to hacking when used correctly.

Reviewer: ```suggestion
It’s worth mentioning that these libraries have been both peer reviewed and audited to accomplish high standards of security so contracts that depend on them are less susceptible to hacking when used correctly.
```

> At terminal, inside the folder token, install OpenZeppelin libraries with this command:

Reviewer: ```suggestion
In the terminal, inside the folder token, install OpenZeppelin libraries with this command:
```

```shell
npm install @openzeppelin/contracts --save
```

![openzeppelin install](/assets/img/tutorials/create-a-token/image-08.png)

More info: 

[openzeppelin.com/contracts](https://openzeppelin.com/contracts/)

## Install HD wallet provider

To connect to the RSK network, we are going to use a provider that allows us to connect to any network unlocking an account locally. 

> We are going to use @truffle/hdwallet-provider. 

Reviewer: ```suggestion
We are going to use `@truffle/hdwallet-provider`. 
```

It can be used to sign transactions for addresses derived from a 12 or 24 word mnemonic.

> You need to have installed Node >= 7.6.

> At terminal, inside the folder token, install it with this command:

Reviewer: ```suggestion
In the terminal, inside the folder token, install it with this command:
```

```shell
npm install @truffle/hdwallet-provider --save
```

![hd wallet provider install](/assets/img/tutorials/create-a-token/image-09.png)

> This is a large package with many facilities. After a while you will see the message of successful installation.

Reviewer: ```suggestion
This is a large package with many utilities. After a while you will see the message of successful installation.
```

![hd wallet provider successful installation](/assets/img/tutorials/create-a-token/image-10.png)

## Check package.json

`package.json` is a file created by npm with some configurations, including the packages which we installed before using the command `npm init -y`.

> After the installations, I will open the project folder named Register in VSCode and verify the file package.json. Let's take a look at the dependencies in the file:

Reviewer: ```suggestion
After the installation, I will open the project folder named `Register` in VSCode and verify the `package.json` file. Let's take a look at the dependencies in the file:
```

![package.json](/assets/img/tutorials/create-a-token/image-11.png)

# Create a mnemonic

Let's create a mnemonic to generate addresses with it.

We are going to use this web app: 

[iancoleman.io/bip39](https://iancoleman.io/bip39/)

> > It is not recommended to be used for any 'real' wallet because it's not a secure way to generate a private key, but can be used for learning purposes.

Reviewer: ```suggestion
> It is not recommended to be used for any 'real' wallet because it's not a secure way to generate a private key, however we will use this here for learning purposes, and since we're using the Testnet anyway.
```

account
```

The result is your account address. 

In my example, it is `0x9682725a85f85f097ab368555a286618dc982c99`. Copy this.

![account address](/assets/img/tutorials/create-a-token/image-16.png)

# TestNet Faucet

You can get some tR-BTC from 

[faucet.testnet.rsk.co](https://faucet.testnet.rsk.co/)

![faucet.testnet.rsk.co](/assets/img/tutorials/create-a-token/image-17.png)

Enter your wallet address that you copied in the last step, and complete the CAPTCHA.

![Wait a few seconds](/assets/img/tutorials/create-a-token/image-18.png)

Wait a few seconds...

![Received some R-BTCs](/assets/img/tutorials/create-a-token/image-19.png)

You can see the transaction hash:

[0x16bedc1339a8fe59e270b0c6d5175851010bb93d0cf6c4974f1705b9ead7ee6e](https://explorer.testnet.rsk.co/tx/0x16bedc1339a8fe59e270b0c6d5175851010bb93d0cf6c4974f1705b9ead7ee6e)

Now I have 0.05 tR-BTC!

## Check balance

Now, let’s check our balance in the console. Run this command in our truffle console:

```javascript
web3.eth.getBalance(account, (err, res) => console.log(res))
```

![getBalance](/assets/img/tutorials/create-a-token/image-20.png)

# Create the smart contract

> In the Contracts folder, create a new file named `Token.sol`

Reviewer: ```suggestion
In the `contracts` folder, create a new file named `Token.sol`.
```

![create Token.sol](/assets/img/tutorials/create-a-token/image-21.png)

# Token.sol with only 7 lines!

This smart contract is a mintable ERC20 token.

Copy and paste the following code.

```javascript
pragma solidity 0.5.2;
import '@openzeppelin/contracts/token/ERC20/ERC20Mintable.sol';
contract Token is ERC20Mintable {
       string public name = "My RSK token";
       string public symbol = "MRT";
       uint8 public decimals = 2;
}
```

> Let me explain the above code.

Reviewer: ```suggestion
Let's review the above code.
```

To create our ERC20 Token, we will import `ERC20Mintable` from Open Zeppelin. 
This library itself imports several other libraries such as `SafeMath.sol`, the standards for this kind of token, and the capability to mint tokens.

Inside the token, we define some basic information about the token: `name`, `symbol`, and number of `decimals` for the precision.

To inherit the library's attributes and functions, we simply define our contract as a `ERC20Mintable` using the `is` keyword in this way.

![Token.sol](/assets/img/tutorials/create-a-token/image-22.png)

# Compile a smart contract

In the truffle console, run this command:

```
compile
```

![truffle compile](/assets/img/tutorials/create-a-token/image-23.png)

# Deploy a smart contract

> First of all, we need to create a file in truffle structure with instructions to deploy the smart contract.

Reviewer: ```suggestion
First of all, we need to create a a new migrations file where truffle will find it, containing instructions to deploy the smart contract.
```

## Create file 2_deploy_contracts.js

> Folder migrations has JavaScript files that help you deploy contracts to the network. 
> These files are responsible for staging your deployment tasks, and they're written under the assumption that your deployment needs will change over time. 
> A history of previously run migrations is recorded on-chain through a special Migrations contract, detailed below. (source: [truffle: running-migrations](https://www.trufflesuite.com/docs/truffle/getting-started/running-migrations))
> 
> At folder migrations, create the file `2_deploy_contracts.js`

Reviewer: ```suggestion
The migrations  folder has JavaScript files that help you deploy contracts to the network. 
These files are responsible for staging your deployment tasks, and they're written under the assumption that your deployment needs will change over time. 
A history of previously run migrations is recorded on-chain through a special Migrations contract, detailed below. (source: [truffle: running-migrations](https://www.trufflesuite.com/docs/truffle/getting-started/running-migrations))

In the migrations folder, create the file `2_deploy_contracts.js`
```

module.exports = function(deployer) {
  deployer.deploy(Token);
};
```

![2_deploy_contracts.js](/assets/img/tutorials/create-a-token/image-25.png)

## Migrate

In the truffle console, run this command:

```
migrate
```

Wait a few minutes while the transactions for the smart contract deployments are sent to the blockchain…

The migrate command will compile the smart contract again if necessary.

![truffle migrate](/assets/img/tutorials/create-a-token/image-26.png)

First, it deploys the smart contract `Migrations.sol`, file generated by Truffle:

![deploy Migrations.sol](/assets/img/tutorials/create-a-token/image-27.png)

This is the transaction:

[0xd29d03fc2b904545005ab6ed205f970575aef184ebecf14c9f0f6b6f45ec1bb3](https://explorer.testnet.rsk.co/tx/0xd29d03fc2b904545005ab6ed205f970575aef184ebecf14c9f0f6b6f45ec1bb3)

And then it deploys our smart contract `Token.sol`:

![deploy Token.sol](/assets/img/tutorials/create-a-token/image-28.png)

This is the transaction:

[0xbfff7cf431bb4af9e1b059dbd6eea935d7d20e52a770c467f38b97b479ba414a](https://explorer.testnet.rsk.co/tx/0xbfff7cf431bb4af9e1b059dbd6eea935d7d20e52a770c467f38b97b479ba414a)

Congratulations!

> `My RSK Token` is published at RSK Testnet.

Reviewer: ```suggestion
`My RSK Token` is now published on the RSK Testnet.
```
